src/algorithms/marginal_algorithm.cc

```diff
@@ -3,70 +3,46 @@
 #include <stan/math/prim/fun.hpp>
 
-#include "lib/progressbar/progressbar.h"
-#include "marginal_state.pb.h"
-#include "src/algorithms/base_algorithm.h"
+#include "algorithm_state.pb.h"
+#include "base_algorithm.h"
 #include "src/collectors/base_collector.h"
 #include "src/mixings/marginal_mixing.h"
-#include "src/utils/eigen_utils.h"
 
 void MarginalAlgorithm::initialize() {
   BaseAlgorithm::initialize();
   marg_mixing = std::dynamic_pointer_cast<MarginalMixing>(mixing);
 }
 
-//! \param grid      Grid of points in matrix form to evaluate the density on
-//! \param collector Collector containing the algorithm chain
-//! \return          Matrix whose i-th column is the lpdf at i-th iteration
-Eigen::MatrixXd MarginalAlgorithm::eval_lpdf(
-    BaseCollector *const collector, const Eigen::MatrixXd &grid,
-    const Eigen::MatrixXd &hier_covariates /*= Eigen::MatrixXd(0, 0)*/,
-    const Eigen::MatrixXd &mix_covariates /*= Eigen::MatrixXd(0, 0)*/) {
-  std::deque<Eigen::VectorXd> lpdf;
-  bool keep = true;
-  progresscpp::ProgressBar bar(collector->get_size(), 60);
-  while (keep) {
-    keep = update_state_from_collector(collector);
-    if (!keep) {
-      break;
-    }
-    lpdf.push_back(lpdf_from_state(grid, hier_covariates, mix_covariates));
-    ++bar;
-    bar.display();
-  }
-  collector->reset();
-  bar.done();
-  return bayesmix::stack_vectors(lpdf);
-}
-
 Eigen::VectorXd MarginalAlgorithm::lpdf_from_state(
     const Eigen::MatrixXd &grid, const Eigen::MatrixXd &hier_covariates,
     const Eigen::MatrixXd &mix_covariates) {
-  Eigen::VectorXd lpdf(grid.rows());
+  // Read mixing state
   unsigned int n_data = curr_state.cluster_allocs_size();
   unsigned int n_clust = curr_state.cluster_states_size();
   marg_mixing->set_state_from_proto(curr_state.mixing_state());
+  // Initialize estimate containers
   Eigen::MatrixXd lpdf_local(grid.rows(), n_clust + 1);
+  Eigen::VectorXd lpdf_final(grid.rows());
   auto temp_hier = unique_values[0]->clone();
-  for (size_t j = 0; j < n_clust; j++) {
-    temp_hier->set_state_from_proto(curr_state.cluster_states(j));
-    lpdf_local.col(j) =
-        marg_mixing->mass_existing_cluster(n_data, true, false, temp_hier) +
-        temp_hier->like_lpdf_grid(grid, hier_covariates).array();
-    // TODO add mixing covariate
-  }
-  lpdf_local.col(n_clust) =
-      marg_mixing->mass_new_cluster(n_data, true, false, n_clust) +
-      lpdf_marginal_component(temp_hier, grid, hier_covariates).array();
-  // TODO add mixing covariate
-
-  for (size_t j = 0; j < grid.rows(); j++) {
-    lpdf(j) = stan::math::log_sum_exp(lpdf_local.row(j));
+  // Loop over grid points
+  for (size_t i = 0; i < grid.rows(); i++) {
+    // Get mass values for i-th grid point
+    double mass_ex = marg_mixing->mass_existing_cluster(
+        n_data, true, false, temp_hier, mix_covariates.row(i));
+    double mass_new = marg_mixing->mass_new_cluster(
+        n_data, true, false, n_clust, mix_covariates.row(i));
```

Comment:
there is something funny going on here: `grid` and `hier_covariates` are different objects with different meanings.
So the main use of this function is the following: suppose you have a single observation with some `hier_covariates` and `mix_covariates` (in this case two vectors) and you want to evaluate the predictive distribution over a grid of values.

For instance you have run the MCMC algorithm and then you have a test-set and want to evaluate the predictive lpdf for a single observation of the test set over the grid (-10, ..., 10).

Here instead you're assuming that for each point of the grid you have a different value of the covariates, which is weird...

Reply:
`grid` and the covariates matrices are scanned in parallel, just like in the main algorithm. If the "single point with several possible covariates" case is common enough, we could create an additional overloaded version which only accepts one single "grid" point!

Reply:
This is different from the main algorithm. In the main algorithm you have observations y_1, ... y_n and covariates x_1, ... x_n, so that they are rightfully scanned in parallel.

Here you want to compute the predictive density on a grid. So basically you need to compute the law of (y* | x*), but you do not observe y*, only x* (e.g. a test set).
Hence, this method returns the evaluation of the density of y* (call it f_x*) on a grid, i.e. given a grid=(g_1, ..., g_k) this method must return (f_x*(g1), ... f_x*(g_k)). Then once you have this density you can use it to compute any point estimate such as the mean of this density or the median / mode, whatever.

Reply:
Ok, gotcha. I will fix it so that you only have two single covariate vectors.